### ISSUE TITLE
porting to python 3

### ISSUE THREAD

**Issue by domeniconappo** [ec-jrc/lisflood-code#28]
1. it will be python 2/3 compatible code
2. Mid 2020: python2 compatible code will be removed and only python3 will be supported

**Comment by domeniconappo** [comment]
Weird results after first rewrite:

tox tests
python 2 env [OK] 
python 3 env differences in results: need to find reason. 

First investigation: **pcraster** python3 version. It seems to behave in a different way because of multicore module...to check if it's true we will force to use old single core functions (i.e. `pcraster.operations.*` instead of `pcraster.multicore._operations.*`

py27:
```console
[OK] dis 3653
[OK] dis 3654
[OK] dis 3655
[OK] dis 3656
[OK] dis 3657
[OK] dis 3658
[OK] dis 3659
[OK] dis 3660
[OK] dis 3661
[OK] dis 3662
[OK] dis 3663
[OK] dis 3664
[OK] dis 3665
[OK] dis 3666
[OK] dis 3667
```

py36:
```console
[OK] dis 3653
[OK] dis 3654
[OK] dis 3655
[OK] dis 3656
[OK] dis 3657
[OK] dis 3658
[OK] dis 3659
[OK] dis 3660
[ERROR]
Var: dis - STEP 3661: 0.280112045% of values are different. max diff: 2.2606
[ERROR]
Var: dis - STEP 3662: 0.280112045% of values are different. max diff: 1.8059
[ERROR]
Var: dis - STEP 3663: 2.240896359% of values are different. max diff: 0.7926
[ERROR]
Var: dis - STEP 3664: 3.781512605% of values are different. max diff: 0.8874
[ERROR]
Var: dis - STEP 3665: 3.781512605% of values are different. max diff: 0.7633
[ERROR]
Var: dis - STEP 3666: 3.781512605% of values are different. max diff: 0.6350
[ERROR]
Var: dis - STEP 3667: 3.781512605% of values are different. max diff: 3.2695

```

**Comment by domeniconappo** [comment]
UPDATE: issue lies in different 'division' behavior between Python2 and Python3 and PCRaster div operator implementations. Just added from __future__ import division to all hydrological modules (in order to have a consistent py3 behavior also in py2 env) and differences pop out in py27 tests too.

**Comment by domeniconappo** [comment]
Found the culprit: reservoir.py module introduced differences by using Py3 division

**Comment by domeniconappo** [comment]
@AdDeRoo @emiliano-gelati @ecCinziaMazzetti @valeriolorini @salampe 

Hi all,
during development on this branch I found out a (serious?) bug in lisflood. 

https://github.com/ec-jrc/lisflood-code/blob/master/src/lisflood/hydrological_modules/reservoir.py#L179

This bug was there since ever, I guess. 
Python2 division between integers returns integers so:

`InvDtSecDay = 1/86400` is 0 and not `1.15740741E-5` ...which is erroneous to the following line

https://github.com/ec-jrc/lisflood-code/blob/master/src/lisflood/hydrological_modules/reservoir.py#L202

Indeed, this is the only thing that leads to differences in Drina test, which seems quite considerable to me...

```
[OK] dis 3653
[OK] dis 3654
[OK] dis 3655
[OK] dis 3656
[OK] dis 3657
[OK] dis 3658
[OK] dis 3659
[OK] dis 3660
[ERROR]
Var: dis - STEP 3661: 0.280112045% of values are different. max diff: 2.2606
[ERROR]
Var: dis - STEP 3662: 0.280112045% of values are different. max diff: 1.8059
[ERROR]
Var: dis - STEP 3663: 2.240896359% of values are different. max diff: 0.7926
[ERROR]
Var: dis - STEP 3664: 3.781512605% of values are different. max diff: 0.8874
[ERROR]
Var: dis - STEP 3665: 3.781512605% of values are different. max diff: 0.7633
[ERROR]
Var: dis - STEP 3666: 3.781512605% of values are different. max diff: 0.6350
[ERROR]
Var: dis - STEP 3667: 3.781512605% of values are different. max diff: 3.2695
[ERROR]
Var: dis - STEP 3668: 3.781512605% of values are different. max diff: 4.8519
[ERROR]
Var: dis - STEP 3669: 3.641456583% of values are different. max diff: 4.9097
[ERROR]
Var: dis - STEP 3670: 6.582633053% of values are different. max diff: 6.0839
[ERROR]
Var: dis - STEP 3671: 6.862745098% of values are different. max diff: 17.9708
[ERROR]
Var: dis - STEP 3672: 6.862745098% of values are different. max diff: 17.5571
[ERROR]
Var: dis - STEP 3673: 6.862745098% of values are different. max diff: 17.7121
[ERROR]
Var: dis - STEP 3674: 6.862745098% of values are different. max diff: 17.8017
[ERROR]
Var: dis - STEP 3675: 6.862745098% of values are different. max diff: 18.0745
[ERROR]
Var: dis - STEP 3676: 6.722689076% of values are different. max diff: 18.0935
[ERROR]
Var: dis - STEP 3677: 6.862745098% of values are different. max diff: 18.1685
[ERROR]
Var: dis - STEP 3678: 6.862745098% of values are different. max diff: 18.7475
[ERROR]
Var: dis - STEP 3679: 6.582633053% of values are different. max diff: 18.2574
[ERROR]
Var: dis - STEP 3680: 6.862745098% of values are different. max diff: 18.7492
[ERROR]
Var: dis - STEP 3681: 6.862745098% of values are different. max diff: 15.2183
[ERROR]
Var: dis - STEP 3682: 6.862745098% of values are different. max diff: 10.0284
[ERROR]
Var: dis - STEP 3683: 6.862745098% of values are different. max diff: 9.7726
[ERROR]
Var: dis - STEP 3684: 6.862745098% of values are different. max diff: 16.0046
[ERROR]
Var: dis - STEP 3685: 6.582633053% of values are different. max diff: 18.6986
[ERROR]
Var: dis - STEP 3686: 6.862745098% of values are different. max diff: 7.4773
[ERROR]
Var: dis - STEP 3687: 4.901960784% of values are different. max diff: 6.6625
[ERROR]
Var: dis - STEP 3688: 3.081232493% of values are different. max diff: 5.8937
[ERROR]
Var: dis - STEP 3689: 3.361344538% of values are different. max diff: 5.2541
[ERROR]
Var: dis - STEP 3690: 3.081232493% of values are different. max diff: 4.7294
[ERROR]
Var: dis - STEP 3691: 3.081232493% of values are different. max diff: 4.2919
[ERROR]
Var: dis - STEP 3692: 3.081232493% of values are different. max diff: 3.9213
[ERROR]
Var: dis - STEP 3693: 5.042016807% of values are different. max diff: 9.9307
[ERROR]
Var: dis - STEP 3694: 5.042016807% of values are different. max diff: 16.1361
[ERROR]
Var: dis - STEP 3695: 3.081232493% of values are different. max diff: 3.0617
[ERROR]
Var: dis - STEP 3696: 3.081232493% of values are different. max diff: 2.8252
[ERROR]
Var: dis - STEP 3697: 3.081232493% of values are different. max diff: 2.6118
[ERROR]
Var: dis - STEP 3698: 3.081232493% of values are different. max diff: 2.4190
[ERROR]
Var: dis - STEP 3699: 3.081232493% of values are different. max diff: 2.2449
[ERROR]
Var: dis - STEP 3700: 3.081232493% of values are different. max diff: 2.0879
[ERROR]
Var: dis - STEP 3701: 3.081232493% of values are different. max diff: 1.9424
[ERROR]
Var: dis - STEP 3702: 3.081232493% of values are different. max diff: 1.8020
[ERROR]
Var: dis - STEP 3703: 3.081232493% of values are different. max diff: 1.6734
[ERROR]
Var: dis - STEP 3704: 3.081232493% of values are different. max diff: 1.5563
F
```

So, to summarise:

LISFLOOD Python3 version (under development) should be the correct one.
I will fix the code and replace the reference with current output.

Please, let me know if there is anything else to consider for this change.

**Comment by emiliano-gelati** [comment]
Hi @domeniconappo,
@ecCinziaMazzetti, @valeriolorini, @salampe, @AdDeRoo,

Thanks for finding the bug and fixing it. Your changes look OK to me. The division in Python2 should be `InvDtSecDay = 1 / 86400.` or `InvDtSecDay = 1 / float(86400)` (the latter may be safer).

I guess the bug was introduced when merging the sub-daily version.

It does not affect the D2 development version, where  `self.var.InvDtSec` is still used to run daily time steps:
https://webgate.ec.europa.eu/CITnet/stash/projects/LISFLOODPY/repos/lisflood/browse/Lisflood/hydrological_modules/reservoir.py#187

**Comment by domeniconappo** [comment]
@emiliano-gelati 

thanks for reviewing!
I'm going to merge with master within end of the week.

Looking forward to merge your D2 version :)
